src/signals-unix.c

```diff
@@ -46,17 +46,89 @@ static bt_context_t *jl_to_bt_context(void *sigctx)
 #endif
 }
 
-static void JL_NORETURN jl_throw_in_ctx(jl_value_t *e, void *sigctx)
+static int thread0_exit_count = 0;
+
+static uintptr_t jl_get_rsp_from_ctx(void *_ctx)
+{
+#if defined(_OS_LINUX_) && defined(_CPU_X86_64_)
+    ucontext_t *ctx = (ucontext_t*)_ctx;
+    return ctx->uc_mcontext.gregs[REG_RSP];
+#elif defined(_OS_LINUX_) && defined(_CPU_X86_)
+    ucontext_t *ctx = (ucontext_t*)_ctx;
+    return ctx->uc_mcontext.gregs[REG_ESP];
+#elif defined(_OS_LINUX_) && defined(_CPU_AARCH64_)
+    ucontext_t *ctx = (ucontext_t*)_ctx;
+    return ctx->uc_mcontext.sp;
+#elif defined(_OS_LINUX_) && defined(_CPU_ARM_)
+    ucontext_t *ctx = (ucontext_t*)_ctx;
+    return ctx->uc_mcontext.arm_sp;
+#elif defined(_OS_DARWIN_)
+    x86_thread_state64_t *state = &((ucontext64_t*)_ctx)->uc_mcontext64->__ss;
+    return state->__rsp;
+#else
+    // TODO Add support for FreeBSD and PowerPC(64)?
+    return 0;
+#endif
+}
+
+static void jl_call_in_ctx(void (*fptr)(void), void *ctx)
+{
+    // Modifying the ucontext should work too but sigreturn orientated
+    // programming mitigation can works against us by rejecting ucontext
+    // that is modified. Set the register values with assembly instead.
+    // This sould be as defined as doing longjmp/siglongjmp/modifying ucontext
+    // on any platforms we care about, which is not really portable but works
+    // in practice. (FWIW this works better than modifying ucontext on Mac).
+    uintptr_t rsp = jl_get_rsp_from_ctx(ctx);
+    if (!rsp) {
+        fptr();
+        return;
+    }
+    // Align the stack and leave 128bytes red zone on all platforms.
+    // (Only a few actually needs these but doesn't hurt to do on all of them).
+    rsp &= -16; // ensure 16-byte alignment
+    rsp -= 128; // 128bytes red zone
+#if defined(_CPU_X86_64_)
```

Comment:
Well, this is in ...-**unix**.c, I'll be happy to learn how to write this in msvc compatible way when we support using MSVC to build for unix =)